test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java

```diff
@@ -359,12 +374,218 @@ public void shouldRejectLetSelectWithIncompletePartitionKey()
                   .hasMessageContaining(String.format(ILLEGAL_RANGE_QUERY_MESSAGE, "LET assignment row1", "at [2:15]"));
     }
 
+    @Test
+    public void shouldRejectSelectDefinedOnTopLevelAndInBranches()
+    {
+        String query = "BEGIN TRANSACTION\n" +
+                       "  LET a = (SELECT * FROM ks.tbl1 WHERE k = 0 AND c = 0);\n" +
+                       "  LET b = (SELECT * FROM ks.tbl1 WHERE k = 1 AND c = 0);\n" +
+                       "  SELECT a.v;\n" +
+                       "  IF a.v = 1 THEN\n" +
+                       "    SELECT b.v;\n" +
+                       "    UPDATE ks.tbl1 SET v = a.v WHERE k=1 AND c=2;\n" +
+                       "  ELSE IF a.v = 2 THEN\n" +
+                       "    SELECT a.v;\n" +
+                       "    UPDATE ks.tbl1 SET v = b.v WHERE k=1 AND c=2;\n" +
+                       "  ELSE\n" +
+                       "    SELECT b.v;\n" +
+                       "  END IF\n" +
+                       "COMMIT TRANSACTION";
+
+        Assertions.assertThatThrownBy(() -> prepare(query))
+                    .isInstanceOf(InvalidRequestException.class)
+                    .hasMessageContaining(String.format(SELECT_ALREADY_DEFINED_MESSAGE, "at [5:3]"));
+    }
+
+    @Test
+    public void shouldRejectSelectDefinedAfterUpdate()
+    {
+        String query = "BEGIN TRANSACTION\n" +
+                       "  LET a = (SELECT * FROM ks.tbl1 WHERE k = 0 AND c = 0);\n" +
+                       "  LET b = (SELECT * FROM ks.tbl1 WHERE k = 1 AND c = 0);\n" +
+                       "  IF a.v = 1 THEN\n" +
+                       "    UPDATE ks.tbl1 SET v = a.v WHERE k=1 AND c=2;\n" +
+                       "    SELECT b.value, a.value;\n" +   // <-- SELECT defined after UPDATE
+                       "  ELSE IF a.v = 2 THEN\n" +
+                       "    SELECT a.value, b.value;\n" +
+                       "    UPDATE ks.tbl1 SET v = b.v WHERE k=1 AND c=2;\n" +
+                       "  ELSE\n" +
+                       "    SELECT a.value, b.value;\n" +
+                       "  END IF\n" +
+                       "COMMIT TRANSACTION";
+
+        Assertions.assertThatThrownBy(() -> prepare(query))
+                  .isInstanceOf(SyntaxException.class)
+                  .hasMessageContaining("mismatched input 'SELECT' expecting K_END");
+    }
+
+    @Test
+    public void shouldRejectMissingSelectInBranch()
```

Comment:
why is this an error? feels weird for me, would expect empty result set in this case rather than an error